patches/server/1046-Async-StoredUserList-saving.patch

```diff
@@ -0,0 +1,192 @@
+index 91feb12732564c90656da487664dbc12e55397fc..6e03808ea6b10695b071f66e5db659ea8bda8e92 100644
+--- a/src/main/java/net/minecraft/server/players/PlayerList.java
++++ b/src/main/java/net/minecraft/server/players/PlayerList.java
+@@ -1333,6 +1333,11 @@ public abstract class PlayerList {
+ 
+         return list;
+     }
++    // Paper start - async StoredUserList saving
++    public void shutdownUserListExecutor() {
++        StoredUserList.shutdownExecutor();
++    }
++    // Paper end
+ 
+     public int getViewDistance() {
+         return this.viewDistance;
+diff --git a/src/main/java/net/minecraft/server/players/StoredUserList.java b/src/main/java/net/minecraft/server/players/StoredUserList.java
+index 35f973cc2c0989256fa21abaf0327c2f36dbe4c9..ab3282b3efb4fea2aa1fb8dcd46a473666f23a37 100644
+--- a/src/main/java/net/minecraft/server/players/StoredUserList.java
++++ b/src/main/java/net/minecraft/server/players/StoredUserList.java
+@@ -32,6 +32,13 @@ public abstract class StoredUserList<K, V extends StoredUserEntry<K>> {
+     private static final Gson GSON = (new GsonBuilder()).setPrettyPrinting().create();
+     private final File file;
+     private final Map<String, V> map = Maps.newConcurrentMap(); // Paper - Use ConcurrentHashMap in JsonList
++    // Paper start - async StoredUserList saving
++    private static final java.util.concurrent.ExecutorService SAVE_EXECUTOR = java.util.concurrent.Executors.newSingleThreadExecutor(
++        new com.google.common.util.concurrent.ThreadFactoryBuilder()
++            .setNameFormat("Async StoredUserList Saving - %1$d")
++            .build()
++    );
++    // Paper end
+ 
+     public StoredUserList(File file) {
+         this.file = file;
+@@ -45,7 +52,7 @@ public abstract class StoredUserList<K, V extends StoredUserEntry<K>> {
+         this.map.put(this.getKeyForUser(entry.getUser()), entry);
+ 
+         try {
+-            this.save();
++            this.save(true); // Paper - async StoredUserList saving
+         } catch (IOException ioexception) {
+             StoredUserList.LOGGER.warn("Could not save the list after adding a user.", ioexception);
+         }
+@@ -65,7 +72,7 @@ public abstract class StoredUserList<K, V extends StoredUserEntry<K>> {
+         this.map.remove(this.getKeyForUser(key));
+ 
+         try {
+-            this.save();
++            this.save(true); // Paper - async StoredUserList saving
+         } catch (IOException ioexception) {
+             StoredUserList.LOGGER.warn("Could not save the list after removing a user.", ioexception);
+         }
+@@ -102,8 +109,14 @@ public abstract class StoredUserList<K, V extends StoredUserEntry<K>> {
+         return this.map.values();
+     }
+ 
+-    public void save() throws IOException {
++    // Paper start - async StoredUserList saving
++    static void shutdownExecutor() {
++        StoredUserList.SAVE_EXECUTOR.shutdown();
```

Comment:
shutdown() does not await running tasks, we'll need to look into e.g. awaitTermination with a *plausible* timeout.

@electronicboy suggests waiting for 30 seconds.

Reply:
done.